src/main/java/net/fabricmc/loader/metadata/ModMetadataParser.java

```diff
@@ -16,66 +16,117 @@
-		.registerTypeAdapter(ModMetadataV0.Dependency.class, new ModMetadataV0.Dependency.Deserializer())
-		.registerTypeAdapter(ModMetadataV0.Person.class, new ModMetadataV0.Person.Deserializer())
-		.create();
-
-	private static final JsonParser JSON_PARSER = new JsonParser();
-
-	private static LoaderModMetadata getMod(FabricLoader loader, JsonObject object) {
-		if (!object.has("schemaVersion")) {
-			return GSON_V0.fromJson(object, ModMetadataV0.class);
-		} else {
-			//noinspection SwitchStatementWithTooFewBranches
-			switch (object.get("schemaVersion").getAsInt()) {
-				case 1:
-					return GSON_V1.fromJson(object, ModMetadataV1.class);
-				default:
-					loader.getLogger().warn("Mod ID " + (object.has("id") ? object.get("id").getAsString() : "<unknown>") + " has invalid schema version: " + object.get("schemaVersion").getAsInt());
-					return null;
+	// Per the ECMA-404 (www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf), the JSON spec does not prohibit duplicate keys.
+	// For all intensive purposes of replicating the logic of Gson before we have migrated to Nanojson, duplicate keys will replace previous entries.
+	public static LoaderModMetadata parseMetadata(Path modJson) throws IOException, ParseMetadataException {
+		try {
+			// So some context:
+			// Per the json specification, ordering of fields is not typically enforced.
+			// Furthermore we cannot guarantee the `schemaVersion` is the first field in every `fabric.mod.json`
+			//
+			// To work around this, we do the following:
+			// Try to read first field
+			// If the first field is the schemaVersion, read the file normally.
+			//
+			// If the first field is not the schema version, fallback to a more exhaustive check.
+			// Read the rest of the file, looking for the `schemaVersion` field.
+			// If we find the field, cache the value
+			// If there happens to be another `schemaVersion` that has a differing value, then fail.
+			// At the end, if we find no `schemaVersion` then assume the `schemaVersion` is 0
+			// Re-read the JSON file.
+			int schemaVersion = 0;
+
+			try (InputStream stream = Files.newInputStream(modJson)) {
+				try (JsonReader reader = new JsonReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+					if (reader.peek() != JsonToken.BEGIN_OBJECT) {
+						throw new ParseMetadataException("Root of \"fabric.mod.json\" must be an object", reader);
+					}
+
+					reader.beginObject();
+
+					boolean firstField = true;
+
+					while (reader.hasNext()) {
+						// Try to read the schemaVersion
+						if (reader.nextName().equals("schemaVersion")) {
+							if (reader.peek() != JsonToken.NUMBER) {
+								throw new ParseMetadataException("\"schemaVersion\" must be a number.", reader);
+							}
+
+							schemaVersion = reader.nextInt();
+
+							if (firstField) {
+								// Finish reading the metadata
+								return ModMetadataParser.readModMetadata(reader, schemaVersion);
```

Comment:
no qualified references to the current class' static members please, it is confusing